### ISSUE TITLE
Intersection Types can be cast to any type V as long as it is nullable

### ISSUE THREAD

**Issue by agrattaqrypt** [microsoft/TypeScript#45640]
# Bug Report

### 🔎 Search Terms

intersection
type
cast
null

### 🕗 Version & Regression Information

- This is a false-negative type checking issue. An error should be generated, but is not.
- This changed between versions 4.2.3 and 4.3.5 and 4.4.2
- The incorrect behavior occurs in 4.3.5. The correct behavior occurs in  4.2.3 and 4.4.2

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.3.5#code/JYOwLgpgTgZghgYwgAgILIN4Chm+XALmQGcwpQBzAbiwF8stRJZEUAhTHPAIyNPJDU6DBAHsQpZKO4ArVEXQBeTPiIAiABYQANttFrk9MRLBTZbIh2UZkvZJp16DR8ZOlyLaZADJkVlYT2Wrr6ADS26gDuolDaACbOIq6mCHCkSmZy+MTIIACuulTIAPTF+NqRcACeOXAgyNBQMVjGkqmk-u4cabkF2kWl5ZU1+PWNza0paWConbKz2b2FJWXjUFL1ACpVAA4QAMoI5DumACwAdABM5wDMo3HIFxeX4SCipnUNUE3r4sjbe0OxzOt3OAFZkC1ksh2jM2AB9fiUZDKdwLHpIwTIAA+S36Ky+Pw2-12ByOwBOjyut3uVOer3eo0JMWJALJwKpN3BkMmMOms3h+QAttxoCjMuicsLRetcflloM1qzSUCKSDrnc6g8nlcGR8xt8WX82arKRcuRCgA)

### 💻 Code

```ts
interface A {
    a: string;
}

interface B {
    b: string;
}

const objA: A = { a: "hello" }
const objB: B = { b: "hello" }
const objAB: A & B = { a: "hello", b: "world" }

const castA = objA as null; // always an error
const castB = objB as null; // always an error
const castAB = objAB as null; // error on TypeScript 4.2.3 and 4.4.2, not an error on TypeScript 4.3.5 
const castAB_string = objAB as string | null; // error on TypeScript 4.2.3 and 4.4.2, not an error on TypeScript 4.3.5 
const castAB_number = objAB as number | null; // error on TypeScript 4.2.3 and 4.4.2, not an error on TypeScript 4.3.5 
```

### 🙁 Actual behavior

In v4.3.5, we don't see the expected type error when casting an intersection type to null or null unions.
This issue does not occur in earlier or later versions.

### 🙂 Expected behavior

The type error behavior does not change depending on what version is being used, unless there is a well-documented breaking change.

**Comment by MartinJohns** [comment]
Sounds like it was just a bug in 4.3.5.

**Comment by agrattaqrypt** [comment]
Yes, I would assume this is something that needs to be patched on the 4.3.x minor version only

**Comment by agrattaqrypt** [comment]
It's worth noting that 4.3.5 is the default version used in the latest VS Code, so it is probably widely used

**Comment by MartinJohns** [comment]
> It's worth noting that 4.3.5 is the default version used in the latest VS Code, so it is probably widely used

But you don't compile with the TypeScript version shipped with VSCode. And I'm confident that with the next August release of VSCode they will ship 4.4, as per microsoft/vscode#130601.

**Comment by agrattaqrypt** [comment]
> > It's worth noting that 4.3.5 is the default version used in the latest VS Code, so it is probably widely used
> 
> But you don't compile with the TypeScript version shipped with VSCode. And I'm confident that with the next August release of VSCode they will ship 4.4, as per [microsoft/vscode#130601](https://github.com/microsoft/vscode/issues/130601).

Okay, well regardless of VS Code, there are people who use 4.3.x to build their code. 4.4.x has breaking changes. There should be a way to address this issue without requiring users to take on breaking changes. This is what the patch version is for.

Ultimately I can't make you fix it. I found a serious bug that wasted my time and I wanted to save other people the hassle by writing up a report.

**Comment by andrewbranch** [comment]
We'll monitor comments on this issue, but if it wasn't noticed until a full version later, it's unlikely this warrants a patch.
